File: cs/src/Ice/ThreadPool.cs
// **********************************************************************
//
// Copyright (c) 2003-2009 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

//
// Uncomment these definitions as needed. (We use #define here instead
// of constants to stop the compiler from complaining about unreachable code.)
//
//#define TRACE_REGISTRATION
//#define TRACE_SHUTDOWN
//#define TRACE_MESSAGE
//#define TRACE_EXCEPTION
//#define TRACE_THREAD
//#define TRACE_STACK_TRACE

namespace IceInternal
{

    using System;
    using System.Collections.Generic;
    using System.Diagnostics;
    using System.Threading;

    public delegate void ThreadPoolWorkItem(bool safeThread);

    public sealed class ThreadPool
    {
        public ThreadPool(Instance instance, string prefix, int timeout)
        {
            _instance = instance;
            _destroyed = false;
            _prefix = prefix;
            _timeout = timeout;
            _size = 0;
            _sizeMax = 0;
            _sizeWarn = 0;
            _threadIndex = 0;
            _running = 0;
            _inUse = 0;
            _load = 1.0;
            _serialize = _instance.initializationData().properties.getPropertyAsInt(_prefix + ".Serialize") > 0;

            string programName = _instance.initializationData().properties.getProperty("Ice.ProgramName");
            if(programName.Length > 0)
            {
                _programNamePrefix = programName + "-";
            }
            else
            {
                _programNamePrefix = "";
            }

            //
            // We use just one thread as the default. This is the fastest
            // possible setting, still allows one level of nesting, and
            // doesn't require to make the servants thread safe.
            //
            int size = _instance.initializationData().properties.getPropertyAsIntWithDefault(_prefix + ".Size", 1);
            if(size < 1)
            {
                string s = _prefix + ".Size < 1; Size adjusted to 1";
                _instance.initializationData().logger.warning(s);
                size = 1;
            }

            int sizeMax =
                _instance.initializationData().properties.getPropertyAsIntWithDefault(_prefix + ".SizeMax", size);
            if(sizeMax < size)
            {
                string s = _prefix + ".SizeMax < " + _prefix + ".Size; SizeMax adjusted to Size (" + size + ")";
                _instance.initializationData().logger.warning(s);
                sizeMax = size;
            }

            int sizeWarn = _instance.initializationData().properties.getPropertyAsIntWithDefault(_prefix + ".SizeWarn",
                                                                              sizeMax * 80 / 100);
            if(sizeWarn > sizeMax)
            {
                string s = _prefix + ".SizeWarn > " + _prefix + ".SizeMax; adjusted SizeWarn to SizeMax ("
                    + sizeMax + ")";
                _instance.initializationData().logger.warning(s);
                sizeWarn = sizeMax;
            }

            _size = size;
            _sizeMax = sizeMax;
            _sizeWarn = sizeWarn;

            try
            {
                _threads = new List<WorkerThread>();
                for(int i = 0; i < _size; ++i)
                {
                    WorkerThread thread = new WorkerThread(this, _programNamePrefix + _prefix + "-" + _threadIndex++);
                    _threads.Add(thread);
                    thread.Start();
                    ++_running;
                }
            }
            catch(System.Exception ex)
            {
                string s = "cannot create thread for `" + _prefix + "':\n" + ex;
                _instance.initializationData().logger.error(s);

                destroy();
                joinWithAllThreads();
                throw;
            }
        }

        public void destroy()
        {
            lock(this)
            {
                #if TRACE_SHUTDOWN
                    trace("destroy");
                #endif

                Debug.Assert(!_destroyed);
                _destroyed = true;
                Monitor.PulseAll(this);
            }
        }

        public void execute(ThreadPoolWorkItem workItem)
        {
            lock(this)
            {
                #if TRACE_REGISTRATION
                    trace("adding work item");
                    #if TRACE_STACK_TRACE
                        Console.Error.WriteLine(Environment.StackTrace);
                    #endif
                #endif

                if(_destroyed)
                {
                    throw new Ice.CommunicatorDestroyedException();
                }
                _workItems.AddLast(workItem);
                Monitor.Pulse(this);
            }
        }

        public void finish(Ice.ConnectionI connection)
        {
            lock(this)
            {
                if(_destroyed)
                {
                    throw new Ice.CommunicatorDestroyedException();
                }
                _finished.AddLast(connection);
                Monitor.Pulse(this);
            }
        }

        public void joinWithAllThreads()
        {
            //
            // _threads is immutable after destroy() has been called,
            // therefore no synchronization is needed. (Synchronization
            // wouldn't be possible here anyway, because otherwise the
            // other threads would never terminate.)
            //
            Debug.Assert(_destroyed);
            foreach(WorkerThread thread in _threads)
            {
                thread.Join();
            }
        }

        public string prefix()
        {
            return _prefix;
        }

        public bool
        serialize()
        {
            return _serialize;
        }

        private void run()
        {
            while(true)
            {
                ThreadPoolWorkItem workItem = null;
                Ice.ConnectionI handler = null;
                lock(this)
                {
                    while(_finished.Count == 0 && _workItems.Count == 0 && !_destroyed)
                    {
                        if(!Monitor.Wait(this, _timeout > 0 ? _timeout * 1000 : Timeout.Infinite))
                        {
                            //
                            // Initiate server shutdown.
                            //
                            try
                            {
                                _instance.objectAdapterFactory().shutdown();
                            }
                            catch(Ice.CommunicatorDestroyedException)
                            {
                            }
                            continue;
                        }
                    }

                    Debug.Assert(_finished.Count > 0 || _workItems.Count > 0 || _destroyed);

                    //
                    // There are two possibilities for a notification:
                    //
                    // 1. The thread pool has been destroyed.
                    //
                    // 2. A work item has been scheduled.
                    //

                    if(_finished.Count > 0)
                    {
                        handler = _finished.First.Value;
                        _finished.RemoveFirst();
                    }
                    else if(_workItems.Count > 0)
                    {
                        //
                        // Work items must be executed first even if the thread pool is destroyed.
                        //
                        workItem = _workItems.First.Value;
                        _workItems.RemoveFirst();
                    }
                    else
                    {
                        Debug.Assert(_destroyed);
                        return;
                    }

                    if(_sizeMax > 1)
                    {
                        Debug.Assert(_inUse >= 0);
                        ++_inUse;

                        if(_inUse == _sizeWarn)
                        {
                            string s = "thread pool `" + _prefix + "' is running low on threads\n"
                                       + "Size=" + _size + ", " + "SizeMax=" + _sizeMax + ", "
                                       + "SizeWarn=" + _sizeWarn;
                            _instance.initializationData().logger.warning(s);
                        }

                        Debug.Assert(_inUse <= _running);
                        if(_inUse < _sizeMax && _inUse == _running)
                        {
                            try
                            {
                                WorkerThread thread = new WorkerThread(this, _programNamePrefix + _prefix + "-" +
                                                                       _threadIndex++);
                                thread.Start();
                                _threads.Add(thread);
                                ++_running;
                            }
                            catch(System.Exception ex)
                            {
                                string s = "cannot create thread for `" + _prefix + "':\n" + ex;
                                _instance.initializationData().logger.error(s);
                            }
                        }
                    }
                }

                //
                // Now we are outside the thread synchronization.
                //
                if(workItem != null)
                {
                    try
                    {
                        //
                        // Execute the work item and indicate whether this thread is safe (i.e., will not be reaped).
                        //
                        workItem(_size == _sizeMax);
                    }
                    catch(Ice.LocalException ex)
                    {
                        string s = "exception in `" + _prefix + "' while calling execute():\n" + ex;
                        _instance.initializationData().logger.error(s);
                    }
                }
                else 
                {
                    Debug.Assert(handler != null);
                    try
                    {
                        //
                        // Execute the work item and indicate whether this thread is safe (i.e., will not be reaped).
                        //
                        handler.finished(this);
                    }
                    catch(Ice.LocalException ex)
                    {
                        string s = "exception in `" + _prefix + "' while calling finished():\n" + ex;
                        _instance.initializationData().logger.error(s);
                    }
                }
                
                if(_sizeMax > 1)
                {
                    lock(this)
                    {
                        if(!_destroyed)
                        {
                            if(_size < _sizeMax) // Dynamic thread pool
                            {
                                //
                                // First we reap threads that have been destroyed before.
                                //
                                int sz = _threads.Count;
                                Debug.Assert(_running <= sz);
                                if(_running < sz)
                                {
                                    List<WorkerThread> liveThreads = new List<WorkerThread>();
                                    foreach(WorkerThread thread in _threads)
                                    {
                                        if(!thread.IsAlive())
                                        {
                                            thread.Join();
                                        }
                                        else
                                        {
                                            liveThreads.Add(thread);
                                        }
                                    }
                                    _threads = liveThreads;
                                }

                                //
                                // Now we check if this thread can be destroyed, based
                                // on a load factor.
                                //

                                //
                                // The load factor jumps immediately to the number of
                                // threads that are currently in use, but decays
                                // exponentially if the number of threads in use is
                                // smaller than the load factor. This reflects that we
                                // create threads immediately when they are needed,
                                // but want the number of threads to slowly decline to
                                // the configured minimum.
                                //
                                double inUse = (double)_inUse;
                                if(_load < inUse)
                                {
                                    _load = inUse;
                                }
                                else
                                {
                                    double loadFactor = 0.05; // TODO: Configurable?
                                    double oneMinusLoadFactor = 1 - loadFactor;
                                    _load = _load * oneMinusLoadFactor + inUse * loadFactor;
                                }

                                if(_running > _size)
                                {
                                    int load = (int)(_load + 0.5);
                                    if(load + 1 < _running)
                                    {
                                        Debug.Assert(_inUse > 0);
                                        --_inUse;

                                        Debug.Assert(_running > 0);
                                        --_running;
                                        return;
                                    }
                                }
                            }

                            Debug.Assert(_inUse > 0);
                            --_inUse;
                        }
                    }

                    #if TRACE_THREAD
                        trace("thread " + Thread.CurrentThread.Name + " is active");
                    #endif
                }
            }
        }

/*
 * Commented out because it is unused.
 *
        private void trace(string msg)
        {
            System.Console.Error.WriteLine(_prefix + "(" + Thread.CurrentThread.Name + "): " + msg);
        }
*/

        private Instance _instance;
        private bool _destroyed;
        private readonly string _prefix;
        private readonly string _programNamePrefix;

        private LinkedList<ThreadPoolWorkItem> _workItems = new LinkedList<ThreadPoolWorkItem>();
        private LinkedList<Ice.ConnectionI> _finished = new LinkedList<Ice.ConnectionI>();

        private int _timeout;

        private sealed class WorkerThread
        {
            private ThreadPool _threadPool;

            internal WorkerThread(ThreadPool threadPool, string name)
                : base()
            {
                _threadPool = threadPool;
                _name = name;
            }

            public bool IsAlive()
            {
                return _thread.IsAlive;
            }

            public void Join()
            {
                _thread.Join();
            }

            public void Start()
            {
                _thread = new Thread(new ThreadStart(Run));
                _thread.IsBackground = true;
                _thread.Name = _name;
                _thread.Start();
            }

            public void Run()
            {
                if(_threadPool._instance.initializationData().threadHook != null)
                {
                    _threadPool._instance.initializationData().threadHook.start();
                }

                try
                {
                    _threadPool.run();
                }
                catch(Ice.LocalException ex)
                {
                    string s = "exception in `" + _threadPool._prefix + "' thread " + _thread.Name + ":\n" + ex;
                    _threadPool._instance.initializationData().logger.error(s);
                }
                catch(System.Exception ex)
                {
                    string s = "unknown exception in `" + _threadPool._prefix + "' thread " + _thread.Name + ":\n" + ex;
                    _threadPool._instance.initializationData().logger.error(s);
                }

                if(_threadPool._instance.initializationData().threadHook != null)
                {
                    _threadPool._instance.initializationData().threadHook.stop();
                }

                #if TRACE_THREAD
                    _threadPool.trace("run() terminated");
                #endif
            }

            private string _name;
            private Thread _thread;
        }

        private readonly int _size; // Number of threads that are pre-created.
        private readonly int _sizeMax; // Maximum number of threads.
        private readonly int _sizeWarn; // If _inUse reaches _sizeWarn, a "low on threads" warning will be printed.
        private readonly bool _serialize; // True if requests need to be serialized over the connection.

        private List<WorkerThread> _threads; // All threads, running or not.
        private int _threadIndex; // For assigning thread names.
        private int _running; // Number of running threads.
        private int _inUse; // Number of threads that are currently in use.
        private double _load; // Current load in number of threads.
    }

}

File: java/src/IceInternal/ThreadPool.java
// **********************************************************************
//
// Copyright (c) 2003-2014 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package IceInternal;

public final class ThreadPool
{
    final class ShutdownWorkItem implements ThreadPoolWorkItem
    {
        public void execute(ThreadPoolCurrent current)
        {
            current.ioCompleted();
            try
            {
                _instance.objectAdapterFactory().shutdown();
            }
            catch(Ice.CommunicatorDestroyedException ex)
            {
            }
        }
    }
    
    static final class FinishedWorkItem implements ThreadPoolWorkItem
    {
        public
        FinishedWorkItem(EventHandler handler)
        {
            _handler = handler;
        }

        public void execute(ThreadPoolCurrent current)
        {
            _handler.finished(current);
        }

        private final EventHandler _handler;
    }
    
    static final class JoinThreadWorkItem implements ThreadPoolWorkItem
    {
        public
        JoinThreadWorkItem(EventHandlerThread thread)
        {
            _thread = thread;
        }
        
        public void execute(ThreadPoolCurrent current)
        {
            // No call to ioCompleted, this shouldn't block (and we don't want to cause
            // a new thread to be started).
            _thread.join();
        }
        
        private final EventHandlerThread _thread;
    }
 
    //
    // Exception raised by the thread pool work queue when the thread pool
    // is destroyed.
    //
    static final class DestroyedException extends RuntimeException
    {
    }

    public
    ThreadPool(Instance instance, String prefix, int timeout)
    {
        _instance = instance;
        _destroyed = false;
        _prefix = prefix;
        _selector = new Selector(instance);
        _threadIndex = 0;
        _inUse = 0;
        _inUseIO = 0;
        _promote = true;
        _serialize = _instance.initializationData().properties.getPropertyAsInt(_prefix + ".Serialize") > 0;
        _serverIdleTime = timeout;
        
        Ice.Properties properties = _instance.initializationData().properties;

        String programName = properties.getProperty("Ice.ProgramName");
        if(programName.length() > 0)
        {
            _threadPrefix = programName + "-" + _prefix;
        }
        else
        {
            _threadPrefix = _prefix;
        }

        int nProcessors = Runtime.getRuntime().availableProcessors();

        //
        // We use just one thread as the default. This is the fastest
        // possible setting, still allows one level of nesting, and
        // doesn't require to make the servants thread safe.
        //
        int size = properties.getPropertyAsIntWithDefault(_prefix + ".Size", 1);
        if(size < 1)
        {
            String s = _prefix + ".Size < 1; Size adjusted to 1";
            _instance.initializationData().logger.warning(s);
            size = 1;
        }               

        int sizeMax = properties.getPropertyAsIntWithDefault(_prefix + ".SizeMax", size);
        if(sizeMax == -1)
        {
            sizeMax = nProcessors;
        }
        if(sizeMax < size)
        {
            String s = _prefix + ".SizeMax < " + _prefix + ".Size; SizeMax adjusted to Size (" + size + ")";
            _instance.initializationData().logger.warning(s);
            sizeMax = size;
        }
                
        int sizeWarn = properties.getPropertyAsInt(_prefix + ".SizeWarn");
        if(sizeWarn != 0 && sizeWarn < size)
        {
            String s = _prefix + ".SizeWarn < " + _prefix + ".Size; adjusted SizeWarn to Size (" + size + ")";
            _instance.initializationData().logger.warning(s);
            sizeWarn = size;
        }
        else if(sizeWarn > sizeMax)
        {
            String s = _prefix + ".SizeWarn > " + _prefix + ".SizeMax; adjusted SizeWarn to SizeMax (" + sizeMax + ")";
            _instance.initializationData().logger.warning(s);
            sizeWarn = sizeMax;
        }

        int threadIdleTime = properties.getPropertyAsIntWithDefault(_prefix + ".ThreadIdleTime", 60);
        if(threadIdleTime < 0)
        {
            String s = _prefix + ".ThreadIdleTime < 0; ThreadIdleTime adjusted to 0";
            _instance.initializationData().logger.warning(s);
            threadIdleTime = 0;
        }

        _size = size;
        _sizeMax = sizeMax;
        _sizeWarn = sizeWarn;
        _sizeIO = Math.min(sizeMax, nProcessors);
        _threadIdleTime = threadIdleTime;

        int stackSize = properties.getPropertyAsInt( _prefix + ".StackSize");
        if(stackSize < 0)
        {
            String s = _prefix + ".StackSize < 0; Size adjusted to JRE default";
            _instance.initializationData().logger.warning(s);
            stackSize = 0;
        }
        _stackSize = stackSize;

        boolean hasPriority = properties.getProperty(_prefix + ".ThreadPriority").length() > 0;
        int priority = properties.getPropertyAsInt(_prefix + ".ThreadPriority");
        if(!hasPriority)
        {
            hasPriority = properties.getProperty("Ice.ThreadPriority").length() > 0;
            priority = properties.getPropertyAsInt("Ice.ThreadPriority");
        }
        _hasPriority = hasPriority;
        _priority = priority;

        _workQueue = new ThreadPoolWorkQueue(this, _instance, _selector);

        _nextHandler = _handlers.iterator();
        
        if(_instance.traceLevels().threadPool >= 1)
        {
            String s = "creating " + _prefix + ": Size = " + _size + ", SizeMax = " + _sizeMax + ", SizeWarn = " +
                       _sizeWarn;
            _instance.initializationData().logger.trace(_instance.traceLevels().threadPoolCat, s);
        }
        
        try
        {
            for(int i = 0; i < _size; i++)
            {
                EventHandlerThread thread = new EventHandlerThread(_threadPrefix + "-" + _threadIndex++);
                _threads.add(thread);
                if(_hasPriority)
                {
                    thread.start(_priority);
                }
                else
                {
                    thread.start(java.lang.Thread.NORM_PRIORITY);
                }
            }
        }
        catch(RuntimeException ex)
        {
            String s = "cannot create thread for `" + _prefix + "':\n" + Ex.toString(ex);
            _instance.initializationData().logger.error(s);

            destroy();
            joinWithAllThreads();
            throw ex;
        }
    }

    protected synchronized void
    finalize()
        throws Throwable
    {
        try
        {
            IceUtilInternal.Assert.FinalizerAssert(_destroyed);
        }
        catch(java.lang.Exception ex)
        {
        }
        finally
        {
            super.finalize();
        }
    }

    public synchronized void
    destroy()
    {
        assert(!_destroyed);
        _destroyed = true;
        _workQueue.destroy();
    }

    public synchronized void
    updateObservers()
    {
        for(EventHandlerThread thread : _threads)
        {
            thread.updateObserver();
        }
    }

    public synchronized void
    initialize(EventHandler handler)
    {
        assert(!_destroyed);
        _selector.initialize(handler);
    }

    public void
    register(EventHandler handler, int op)
    {
        update(handler, SocketOperation.None, op);
    }

    public synchronized void
    update(EventHandler handler, int remove, int add)
    {
        assert(!_destroyed);
        _selector.update(handler, remove, add);
    }

    public void
    unregister(EventHandler handler, int op)
    {
        update(handler, op, SocketOperation.None);
    }

    public synchronized void
    finish(EventHandler handler)
    {
        assert(!_destroyed);
        _selector.finish(handler);
        _workQueue.queue(new FinishedWorkItem(handler));
    }

    public void
    execute(ThreadPoolWorkItem workItem)
    {
        _workQueue.queue(workItem);
    }

    public void
    joinWithAllThreads()
    {
        //
        // _threads is immutable after destroy() has been called,
        // therefore no synchronization is needed. (Synchronization
        // wouldn't be possible here anyway, because otherwise the
        // other threads would never terminate.)
        //
        for(EventHandlerThread thread : _threads)
        {
            thread.join();
        }

        //
        // Destroy the selector
        //
        _workQueue.close();
        _selector.destroy();
    }

    public String
    prefix()
    {
        return _prefix;
    }

    private void
    run(EventHandlerThread thread)
    {
        ThreadPoolCurrent current = new ThreadPoolCurrent(_instance, this, thread);
        boolean select = false;
        while(true)
        {
            if(current._handler != null)
            {
                try
                {
                    current._handler.message(current);
                }
                catch(DestroyedException ex)
                {
                    return;
                }
                catch(java.lang.Exception ex)
                {
                    String s = "exception in `" + _prefix + "':\n" + Ex.toString(ex);
                    s += "\nevent handler: " + current._handler.toString();
                    _instance.initializationData().logger.error(s);
                }
            }
            else if(select)
            {
                try
                {
                    _selector.select(_serverIdleTime);
                }
                catch(Selector.TimeoutException ex)
                {
                    synchronized(this)
                    {
                        if(!_destroyed && _inUse == 0)
                        {
                            _workQueue.queue(new ShutdownWorkItem()); // Select timed-out.
                        }
                        continue;
                    }
                }
            }

            synchronized(this)
            {
                if(current._handler == null)
                {
                    if(select)
                    {
                        _selector.finishSelect(_handlers, _serverIdleTime);
                        _nextHandler = _handlers.iterator();
                        select = false;
                    }
                    else if(!current._leader && followerWait(current))
                    {
                        return; // Wait timed-out.
                    }
                }
                else if(_sizeMax > 1)
                {
                    if(!current._ioCompleted)
                    {
                        //
                        // The handler didn't call ioCompleted() so we take care of decreasing 
                        // the IO thread count now.
                        //
                        --_inUseIO;

                        if((current.operation & SocketOperation.Read) != 0 && current._handler.hasMoreData())
                        {
                            _selector.hasMoreData(current._handler);
                        }
                    }
                    else
                    {
                        //
                        // If the handler called ioCompleted(), we re-enable the handler in
                        // case it was disabled and we decrease the number of thread in use.
                        //
                        _selector.enable(current._handler, current.operation);
                        assert(_inUse > 0);
                        --_inUse;
                    }

                    if(!current._leader && followerWait(current))
                    {
                        return; // Wait timed-out.
                    }
                }
                else if(!current._ioCompleted && 
                        (current.operation & SocketOperation.Read) != 0 && current._handler.hasMoreData())
                {
                    _selector.hasMoreData(current._handler);
                }

                //
                // Get the next ready handler.
                //
                if(_nextHandler.hasNext())
                {
                    current._ioCompleted = false;
                    current._handler = _nextHandler.next();
                    current.operation = current._handler._ready;
                    thread.setState(Ice.Instrumentation.ThreadState.ThreadStateInUseForIO);
                }
                else
                {
                    current._handler = null;
                }

                if(current._handler == null)
                {
                    //
                    // If there are no more ready handlers and there are still threads busy performing 
                    // IO, we give up leadership and promote another follower (which will perform the
                    // select() only once all the IOs are completed). Otherwise, if there's no more 
                    // threads peforming IOs, it's time to do another select().
                    //
                    if(_inUseIO > 0)
                    {
                        promoteFollower(current);
                    }
                    else
                    {
                        _selector.startSelect();
                        select = true;
                        thread.setState(Ice.Instrumentation.ThreadState.ThreadStateIdle);
                    }
                }
                else if(_sizeMax > 1)
                {
                    //
                    // Increment the IO thread count and if there's still threads available
                    // to perform IO and more handlers ready, we promote a follower.
                    //
                    ++_inUseIO;
                    if(_nextHandler.hasNext() && _inUseIO < _sizeIO)
                    {
                        promoteFollower(current);
                    }
                }
            }
        }
    }

    synchronized void
    ioCompleted(ThreadPoolCurrent current)
    {
        current._ioCompleted = true; // Set the IO completed flag to specify that ioCompleted() has been called.

        current._thread.setState(Ice.Instrumentation.ThreadState.ThreadStateInUseForUser);

        if(_sizeMax > 1)
        {
            --_inUseIO;
            
            if((current.operation & SocketOperation.Read) != 0 && current._handler.hasMoreData())
            {
                _selector.hasMoreData(current._handler);
            }
            
            if(_serialize && !_destroyed)
            {
                _selector.disable(current._handler, current.operation);
            }    
                
            if(current._leader)
            {
                //
                // If this thread is still the leader, it's time to promote a new leader.
                //
                promoteFollower(current);
            }
            else if(_promote && (_nextHandler.hasNext() || _inUseIO == 0))
            {
                notify();
            }

            assert(_inUse >= 0);
            ++_inUse;
                
            if(_inUse == _sizeWarn)
            {
                String s = "thread pool `" + _prefix + "' is running low on threads\n"
                    + "Size=" + _size + ", " + "SizeMax=" + _sizeMax + ", " + "SizeWarn=" + _sizeWarn;
                _instance.initializationData().logger.warning(s);
            }
                
            if(!_destroyed)
            {            
                assert(_inUse <= _threads.size());
                if(_inUse < _sizeMax && _inUse == _threads.size())
                {
                    if(_instance.traceLevels().threadPool >= 1)
                    {
                        String s = "growing " + _prefix + ": Size=" + (_threads.size() + 1);
                        _instance.initializationData().logger.trace(_instance.traceLevels().threadPoolCat, s);
                    }
                        
                    try
                    {
                        EventHandlerThread thread = new EventHandlerThread(_threadPrefix + "-" + _threadIndex++);
                        _threads.add(thread);
                        if(_hasPriority)
                        {
                            thread.start(_priority);
                        }
                        else
                        {
                            thread.start(java.lang.Thread.NORM_PRIORITY);
                        }
                    }
                    catch(RuntimeException ex)
                    {
                        String s = "cannot create thread for `" + _prefix + "':\n" + Ex.toString(ex);
                        _instance.initializationData().logger.error(s);
                    }
                }
            }
        }
        else if((current.operation & SocketOperation.Read) != 0 && current._handler.hasMoreData())
        {
            _selector.hasMoreData(current._handler);
        }
    }
    
    private synchronized void
    promoteFollower(ThreadPoolCurrent current)
    {
        assert(!_promote && current._leader);
        _promote = true;
        if(_inUseIO < _sizeIO && (_nextHandler.hasNext() || _inUseIO == 0))
        {
            notify();
        }
        current._leader = false;
    }

    private synchronized boolean
    followerWait(ThreadPoolCurrent current)
    {
        assert(!current._leader);

        current._thread.setState(Ice.Instrumentation.ThreadState.ThreadStateIdle);

        //
        // It's important to clear the handler before waiting to make sure that
        // resources for the handler are released now if it's finished. We also
        // clear the per-thread stream.
        //
        current._handler = null;
        current.stream.reset();
        
        //
        // Wait to be promoted and for all the IO threads to be done.
        //
        while(!_promote || _inUseIO == _sizeIO || !_nextHandler.hasNext() && _inUseIO > 0)
        {
            try
            {
                if(_threadIdleTime > 0)
                {
                    long before = IceInternal.Time.currentMonotonicTimeMillis();
                    wait(_threadIdleTime * 1000);
                    if(IceInternal.Time.currentMonotonicTimeMillis() - before >= _threadIdleTime * 1000)
                    {
                        if(!_destroyed && (!_promote || _inUseIO == _sizeIO ||
                                           (!_nextHandler.hasNext() && _inUseIO > 0)))
                        {
                            if(_instance.traceLevels().threadPool >= 1)
                            {
                                String s = "shrinking " + _prefix + ": Size=" + (_threads.size() - 1);
                                _instance.initializationData().logger.trace(_instance.traceLevels().threadPoolCat, s);
                            }
                            assert(_threads.size() > 1); // Can only be called by a waiting follower thread.
                            _threads.remove(current._thread);
                            _workQueue.queue(new JoinThreadWorkItem(current._thread));
                            return true;
                        }
                    }
                }
                else
                {
                    wait();
                }
            }
            catch(InterruptedException ex)
            {
            }
        }
        current._leader = true; // The current thread has become the leader.
        _promote = false;
        return false;
    }

    private final Instance _instance;
    private final ThreadPoolWorkQueue _workQueue;
    private boolean _destroyed;
    private final String _prefix;
    private final String _threadPrefix;
    private final Selector _selector;

    final class EventHandlerThread implements Runnable
    {
        EventHandlerThread(String name)
        {
            _name = name;
            _state = Ice.Instrumentation.ThreadState.ThreadStateIdle;
            updateObserver();
        }

        public void
        updateObserver()
        {
            // Must be called with the thread pool mutex locked
            Ice.Instrumentation.CommunicatorObserver obsv = _instance.getObserver();
            if(obsv != null)
            {
                _observer = obsv.getThreadObserver(_prefix, _name, _state, _observer);
                if(_observer != null)
                {
                    _observer.attach();
                }
            }
        }

        public void
        setState(Ice.Instrumentation.ThreadState s)
        {
            // Must be called with the thread pool mutex locked
            if(_observer != null)
            {
                if(_state != s)
                {
                    _observer.stateChanged(_state, s);
                }
            }
            _state = s;
        }

        public void
        join()
        {
            while(true)
            {
                try
                {
                    _thread.join();
                    break;
                }
                catch(InterruptedException ex)
                {
                }
            }
        }

        public void
        start(int priority)
        {
            _thread = new Thread(null, this, _name, _stackSize);
            _thread.setPriority(priority);
            _thread.start();
        }

        public void
        run()
        {
            if(_instance.initializationData().threadHook != null)
            {
                try
                {
                    _instance.initializationData().threadHook.start();
                }
                catch(java.lang.Exception ex)
                {
                    String s = "thread hook start() method raised an unexpected exception in `";
                    s += _prefix + "' thread " + _name + ":\n" + Ex.toString(ex);
                    _instance.initializationData().logger.error(s);
                }
            }

            try
            {
                ThreadPool.this.run(this);
            }
            catch(java.lang.Exception ex)
            {
                String s = "exception in `" + _prefix + "' thread " + _name + ":\n" + Ex.toString(ex);
                _instance.initializationData().logger.error(s);
            }

            if(_observer != null)
            {
                _observer.detach();
            }

            if(_instance.initializationData().threadHook != null)
            {
                try
                {
                    _instance.initializationData().threadHook.stop();
                }
                catch(java.lang.Exception ex)
                {
                    String s = "thread hook stop() method raised an unexpected exception in `";
                    s += _prefix + "' thread " + _name + ":\n" + Ex.toString(ex);
                    _instance.initializationData().logger.error(s);
                }
            }
        }

        final private String _name;
        private Thread _thread;
        private Ice.Instrumentation.ThreadState _state;
        private Ice.Instrumentation.ThreadObserver _observer;
    }

    private final int _size; // Number of threads that are pre-created.
    private final int _sizeIO; // Number of threads that can concurrently perform IO.
    private final int _sizeMax; // Maximum number of threads.
    private final int _sizeWarn; // If _inUse reaches _sizeWarn, a "low on threads" warning will be printed.
    private final boolean _serialize; // True if requests need to be serialized over the connection.
    private final int _priority;
    private final boolean _hasPriority;
    private final long _serverIdleTime;
    private final long _threadIdleTime;
    private final int _stackSize;

    private java.util.List<EventHandlerThread> _threads = new java.util.ArrayList<EventHandlerThread>();
    private int _threadIndex; // For assigning thread names.
    private int _inUse; // Number of threads that are currently in use.
    private int _inUseIO; // Number of threads that are currently performing IO.

    private java.util.List<EventHandler> _handlers = new java.util.ArrayList<EventHandler>();
    private java.util.Iterator<EventHandler> _nextHandler;

    private boolean _promote;
}
